Core/PixelExperiment.swift

```diff
@@ -0,0 +1,119 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+public enum PixelExperiment: String, CaseIterable {
+
+    fileprivate static let logic = PixelExperimentLogic {
+        Pixel.fire(pixel: $0)
+    }
+
+    /// When `cohort` is accessed for the first time after the experiment is installed with `install()`,
+    ///  allocate and return a cohort.  Subsequently, return the same cohort.
+    public static var cohort: PixelExperiment? {
+        logic.cohort
+    }
+
+    static var isExperimentInstalled: Bool {
+        return logic.isInstalled
+    }
+
+    static var allocatedCohortDoesNotMatchCurrentCohorts: Bool {
+        guard let allocatedCohort = logic.allocatedCohort else { return false }
+        if PixelExperiment(rawValue: allocatedCohort) == nil {
+            return true
+        }
+        return false
+    }
+
+    /// Enables this experiment for new users when called from the new installation path.
+    public static func install() {
+        // Disable the experiment until all other experiments are finished
+        // logic.install()
+    }
+
+    static func cleanup() {
+        logic.cleanup()
+    }
+
+    // These are the variants. Rename or add/remove them as needed.  If you change the string value
+    //  remember to keep it clear for privacy triage.
+    case control
+    case newSettings = "variant1"
+}
+
+extension PixelExperiment {
+
+    // Pixel parameter - cohort
+    public static var parameters: [String: String] {
+        guard let cohort else {
+            return [:]
+        }
+
+        return [PixelParameters.cohort: cohort.rawValue]
+    }
+
+}
+
+final internal class PixelExperimentLogic {
+
+    var cohort: PixelExperiment? {
+        guard isInstalled else { return nil }
+
+        if let allocatedCohort,
+           // if the stored cohort doesn't match, allocate a new one
+           let cohort = PixelExperiment(rawValue: allocatedCohort) {
+            return cohort
+        }
+
+        // For now, just use equal distribution of all cohorts.
+        let cohort = PixelExperiment.allCases.randomElement()!
+        allocatedCohort = cohort.rawValue
+        fireEnrollmentPixel()
+        return cohort
+    }
+
+    @UserDefaultsWrapper(key: .pixelExperimentInstalled, defaultValue: false)
+    var isInstalled: Bool
+
+    @UserDefaultsWrapper(key: .pixelExperimentCohort, defaultValue: nil)
+    var allocatedCohort: String?
+
+    private let fire: (Pixel.Event) -> Void
+
+    init(fire: @escaping (Pixel.Event) -> Void) {
+        self.fire = fire
+    }
+
+    func install() {
+        isInstalled = true
+    }
+
+    func fireEnrollmentPixel() {
+        Pixel.fire(pixel: .pixelExperimentEnrollment,
+                   withAdditionalParameters: PixelExperiment.parameters)
```

Comment:
Was the idea to call the fire function passed in?  In the original macOS code (which I see no longer has this) the idea was to be able to mock calls to pixels so that we could test it.

Talking of which - I don't see any unit tests? I think we should because now it's in the code base there's a good chance this will end up being the defacto standard approach moving forward.

Reply:
Since the experiment uses existing pixels and it adds many pixels on top and few pixels, the current approach made more sense to me.

- Mocking would create approximately 2x30 methods which would be hard to maintain.
- Adding existing pixels to the experiment or removal of a pixel from the experiment would require to swap the standard Pixel.fire call with the specific method.

- Do you think wrapper calls are that important? I don't have a strong opinion and can generate them (uncle GPT can help with that)

Reply:
I see, thanks for clarification! 👍 I missed this part. I have modified it slightly and added couple of tests to ensure pixel is fired when a cohort is allocated. Also, there is a verification the experiment logic returns appropriate pixel parameters.